File: Golang/DynamicProgramming/Subsequence/LongestIncreasingSub.go
package main

func Max(a, b int)int{
	if a > b{
		return a
	}
	return b
}

func lengthOfLIS(nums []int)int{
	n := len(nums)
	if n == 0 {
		return 0
	}
	dp := make([]int, n)
	for i := 0; i < n; i++{
		dp[i] = 1;
	}

	maxLen := 1
	for i := 1; i < n; i++{
		for j := 0; j < i; j++{
			if nums[i] > nums[j]{
				dp[i] = Max(dp[j] + 1, dp[i])
			}
		}
		maxLen = Max(dp[i], maxLen)
	}
	return maxLen
}

File: Java/BinarySearch/LongestIncreasingSub.java
package Java.BinarySearch;

import java.util.ArrayList;
import java.util.List;

public class LongestIncreasingSub {
    public static int lengthOfLIS(int[] nums){
        List<Integer> res = new ArrayList<>();
        for(int n : nums){
            if(res.isEmpty() || res.get(res.size() - 1) < n){
                res.add(n);
            }else{
                int idx = binarySearch(res, n);
                res.set(idx, n);
            }
        }
        return res.size();
    }

    public static int binarySearch(List<Integer> arr, int target){
        int left = 0, right = arr.size() - 1;

        while(left <= right){
            int mid = (left + right) / 2;
            if(arr.get(mid) == target){
                return mid;
            }else if(arr.get(mid) > target){
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return left;
    }
    public static void main(String[] args) {
        
    }
}
